l3kernel/l3prop.dtx

```diff
@@ -49,32 +49,73 @@
 %
 % \begin{documentation}
 %
-% \pkg{expl3} implements a property list data type, which contain
-% an unordered list of entries each of which consists of a \meta{key} and
-% an associated \meta{value}. The \meta{key} and \meta{value} may both
-% be any balanced text, and the \meta{key} is processed using
-% \cs{tl_to_str:n}, meaning that category codes are ignored. It is possible to
-% map functions to property lists such that the function is applied to every
-% key--value pair within the list.
+% \pkg{expl3} implements a \enquote{property list} data type, which contain
+% an unordered list of entries each of which consists of a \meta{key} (string)
+% and an associated \meta{value} (token list). The \meta{key} and \meta{value}
+% may both be given as any balanced text, and the \meta{key} is processed using
+% \cs{tl_to_str:n}, meaning that category codes are ignored. Entries can be
+% manipulated individually, as well as collectively by applying a function to
+% every key--value pair within the list.
 %
 % Each entry in a property list must have a unique \meta{key}: if an entry is
 % added to a property list which already contains the \meta{key} then the new
 % entry overwrites the existing one. The \meta{keys} are compared on a
 % string basis, using the same method as \cs{str_if_eq:nnTF}.
 %
 % Property lists are intended for storing key-based information for use within
-% code.  This is in contrast to key--value lists, which are a form of
+% code. They can be converted from and to key--value lists, which are a form of
 % \emph{input} parsed by the \pkg{l3keys} module.
 %
+% Internally, property lists can use two distinct implementations with different
+% data storage, which are decided when declaring the property list variable
+% using \cs{prop_new:N} (\enquote{flat} storage) or \cs{prop_new_large:N}
+% (\enquote{linked} storage). Once a property list is declared with
+% \cs{prop_new:N} or \cs{prop_new_large:N}, the type of internal data storage
+% can no longer be changed. All other \pkg{l3prop} functions transparently
+% manipulate either storage method and convert as needed.
+% \begin{itemize}
+%   \item
+%     The (default) \enquote{flat} storage method is suited for a relatively
+%     small number of entries, or when the property list is likely to be
+%     manipulated (copied, mapped) as a whole rather than entry-wise.  It is
+%     significantly faster for \cs{prop_set_eq:NN}, and only slightly faster for
+%     \cs{prop_clear:N}, \cs{prop_concat:NNN}, and mapping functions
+%     \cs[no-index]{prop_map_\ldots{}}.
+%
+%   \item
+%     The \enquote{linked} storage method is meant for property lists with a
+%     large numbers of entries.  It has more memory overhead, but is
```

Comment:
Imho "memory overhead" should be clarified. Half of the user will probably think that that means computer memory.